src/FunctionApps/DSTP/src/main/kotlin/gov/cdc/prime/phdi/ConvertVXUToFHIR.kt

```diff
@@ -0,0 +1,81 @@
+            path = "bronze/decrypted/vxu/{name}",
+            connection="AzureWebJobsStorage"
+        ) content: ByteArray,
+        // @BindingName("name") filename: String,
+        @BlobOutput(
+            name="validTarget",
+            dataType = "string",
+            path="bronze/valid-messages/{name}.fhir",
+            connection="AzureWebJobsStorage"
+        ) validContent: OutputBinding<String>,
+        @BlobOutput(
+            name="invalidTarget",
+            dataType = "string",
+            path="bronze/invalid-messages/{name}.txt",
+            connection="AzureWebJobsStorage"
+        ) invalidContent: OutputBinding<String>,
+        context: ExecutionContext
+	) {
+        // no logging until we can confirm that filename is never sensitive
+        //context.logger.info("Name: ${filename} Size: ${content.size} bytes.")
+
+        /* 
+            Every time a new blob is moved to bronze/decrypted/vxu, this function
+            triggers, reads in the data, confirms that it's valid, converts the data
+            to FHIR format, and then stores it in the bronze/valid-messages container. 
+            If the data is invalid, the FHIR conversion is skipped and the data is 
+            stored in bronze/invalid-messages in its original format.
+        */
+        val rawMessages: String = readHL7MessagesFromByteArray(content)
+        val processedMessages: List<String> = convertBatchMessagesToList(rawMessages)
+        val validMessages = StringBuilder()
+        val invalidMessages = StringBuilder()
+        /* TODO: 
+            Figure out if an access token is even necessary. If it is
+            then figure out a way to get the expiration time from the
+            environment variables and only grab a new token if necessary.
+        */    
+        val accessToken: String? = getAccessToken()
+
+        if (!accessToken.isNullOrEmpty()) {
+            processedMessages.forEach {
+                if (isValidHL7Message(it)) {
+                    val json = convertMessageToFHIR(it, "hl7v2", "VXU_V04", accessToken)
+                    if (isValidFHIRMessage(json)) {
+                        validMessages.append("$json\n")
+                    } else {
+                        context.logger.info("An invalid FHIR message was returned during the conversion process.")
+                        invalidMessages.append("$it\n")
+                    }
+                } else {
+                    context.logger.info("Failed to convert a message to FHIR.")
+                    invalidMessages.append("$it\n")
+                }
+            }
+        } else {
+            context.logger.info("Failed to retrieve a valid access token.")
+        }
+
+        if (validMessages.isNotBlank()) {
+            validContent.setValue(validMessages.toString().trim())
+        }
+
+        if (invalidMessages.isNotBlank()) {
+            invalidContent.setValue(invalidMessages.toString().trim())
+        }
+    }
+}
```

Comment:
I was thinking about this, and while we could definitely refactor to minimize repetition, I was wondering if we'd actually benefit from doing so. I think keeping the code inside of the Function itself creates a level of readability with regard to what the Function is doing. It also gives us some flexibility in handling the ELR, ECR, and VXU messages in a custom way. For right now, in the prototype, there is lots of overlap, but maybe that won't be true going forward. Either way, my initial conclusion was that there was no value in further refactoring the code, but I could be convinced to change my mind.

Reply:
I think it's more about reducing maintenance than reducing repetition (fixing something in one but not the other place, having to duplicate functionality, etc), especially since they're untested. I can see things diverging down the line, but it looks like most of the difference at this point is the record type sent to `convertMessageToFHIR`.